### PR TITLE
[AMDGPU][IGLP] SingleWaveOpt: Cache DSW Counters from PreRA

### PR DESCRIPTION
**Pull request by jrbyrnes**
Allowing PERMs to fall into the generic VALU for this mutation is creating edges which block them being classified into the V_PERM SchedGroups. Since other groups in the pipeline depend on V_PERM placement (e.g. DS_WRITE group must be a succ of the V_PERM in stage 2a), not being able to place the V_PERM is a bit catastrophic.

Additionally, remove a bad assert in SharesPredWithPrevNthGroup -- it is possible that we will have an empty cache if OtherGroup has no elements which have a V_PERM pred (possible if the V_PERM SG is empty).

### CODE REVIEW

**Comment by jrbyrnes** [comment]
Latest commit adds "caching" of the DSW counters from PreRA scheduling. While this avoids recalculation in the postRA pass, that isn't the main purpose.

This is required because of physical register dependencies in PostRA scheduling -- they alter the DAG s.t. our counters may become incorrect -- which alters the layout of the pipeline. By preserving the values from PreRA, we can be sure that we accurately construct the pipeline.

Tighter matching on VALU / Perm essentially provided a workaround to cope with this transient behavior. That said, the further constraint on SG matching should only make the mutation more robust.

File: llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp [hunk 1]
```diff
@@ -900,6 +900,11 @@ void MFMASmallGemmOpt::applyIGLPStrategy(
   }
 }
 
+static unsigned DSWCount = 0;
+static unsigned DSWWithPermCount = 0;
+static unsigned DSWWithSharedVMEMCount = 0;
+static bool HasDSWCounts = false;
```
[the remark this answers is not in the document]
Comment (jrbyrnes) [reply llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp]: Thanks for taking a look.

Is this what you had in mind? I've added as a separate commit to (hopefully) making reviewing easier -- they will be squashed before landing.

**Review by kerbowa — APPROVED** [review]
LGTM

File: llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp [hunk 2]
```diff
@@ -1105,14 +1112,20 @@ class MFMASmallGemmSingleWaveOpt final : public IGLPStrategy {
...
+    DenseMap<int, SmallVector<SchedGroup, 4>> &SyncedSchedGroups,
+    bool IsPostRA) {
   unsigned MFMACount = 0;
-  unsigned DSWCount = 0;
-  unsigned DSWWithPermCount = 0;
-  unsigned DSWWithSharedVMEMCount = 0;
   unsigned DSRCount = 0;
+
+  assert((IsPostRA ||
+          DSWCount == DSWWithPermCount == DSWWithSharedVMEMCount == 0) &&
```
Comment (mikaelholmen) [inline llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp:1127]: Hi @jrbyrnes 
I don't know this code at all, but gcc warned on this assert and suggested some paratheses, so it caught my eye.
Does it really check what the assert string says?
I think e.g. the assert passes if the counts are 1, 1, 0 but the string says all should be 0?

Comment (jrbyrnes) [reply llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp:1127]: Yes, you are right. Thanks for pointing that out.

**Pull request merged**